### ISSUE TITLE
Scaling train and test return different results

### ISSUE THREAD

**Issue by shaolang** [elixir-nx/scholar#217]
Should scalers operate on the same scale on the test data after fitting it with the training data?
When I run the following, I (expectedly) get different values for values 1, 2, and 3:

```elixir
train = Nx.tensor([1, 2, 3, 100])
Scholar.Preprocessing.min_max_scale(train)    # returns [0.0, 0.010101, 0.020202, 1.0]

test = Nx.tensor([1, 2, 3])
Scholar.Preprocessing.min_max_scale(test)     # returns [0.0, 0.5, 1.0]
```

The model trained with the scaled training data will perform badly on the test data, 'cos of the scaling difference.

I may be missing something but I think scalers need to the return a model (?) of the scaling operation, so that scaling on test/validation/production data will use the same scale and thus improving the model's performance.

**Comment by josevalim** [comment]
This is a duplicate of #179. /cc @santiago-imelio 

I think we need to revisit Scholar.Preprocessing, because we currently only support the `fit_predict`/`fit_transform` version of said functions. I am in favor of converting some of these functions to modules, such as `Scholar.Preprocessing.StandardScaler`. I know @msluszniak had reservations though.

The main questions are, which functions in `Scholar.Preprocessing` need to become modules? All of them? Or there are some that do not fit the `fit_predict`/`fit_transform` model?

**Comment by santiago-imelio** [comment]
@josevalim @shaolang Hey, I agree that we should start migrating some of the preprocessing functions into modules. Initially we could migrate scalers in order to fill this gap, and later look into other more complex functions like one-hot encoding, what do you think?

**Comment by shaolang** [comment]
@santiago-imelio Sounds good to me. I see that you've already done some work for ticket #179. Have you done anything for min-max scaling too? If not, I'll try contributing with that as reference.

**Comment by josevalim** [comment]
The biggest question is if we should remove `Scholar.Preprocessing.min_max_scale` and friends, or keep it as a delegation to the modules we introduce.

**Comment by shaolang** [comment]
If we keep `Scholar.Preprocessing.min_max_scale` and friends (and delegate to the respective modules' `fit_transform`), I'd imagine their return values will change depending on whether the invocation is for training or otherwise, maybe determined by the presence of a given scaler, e.g.,

```elixir
def min_max_scale(tensor, scaler \\ nil, opts \\ []) do
   # ...

   if is_nil(scaler), do: {scaled_values, new_scaler}, else: scaled_values
end
```

WDYT?

**Comment by josevalim** [comment]
@shaolang can you provide examples of where it would be different? My understanding is that the result would be the same. For min-max scale, wouldn't `fit` be to calculate the bounds and then `transform` applies the bounds? It is what we have today in `min_max_scale`, but broken in steps?

**Comment by santiago-imelio** [comment]
@shaolang Haven't done anything for min-max scaler, so go ahead :)

**Comment by shaolang** [comment]
> @shaolang can you provide examples of where it would be different? My understanding is that the result would be the same. For min-max scale, wouldn't `fit` be to calculate the bounds and then `transform` applies the bounds? It is what we have today in `min_max_scale`, but broken in steps?

@josevalim  Yes, the current one is broken in steps; when training, min-max scale needs to fit the training data to work out the scales. then transform the training data with the scales (aka the current `fit_transform` step you mentioned). When testing/validating, it should skip the fitting step and go straight to transforming; however, the current implementation doesn't accept the "learned" scales, so scaling done during validating/testing produces different results, and hence affecting the model's performance.

> ```elixir
> train = Nx.tensor([1, 2, 3, 100])
> Scholar.Preprocessing.min_max_scale(train)    # returns [0.0, 0.010101, 0.020202, 1.0]
> 
> test = Nx.tensor([1, 2, 3])
> Scholar.Preprocessing.min_max_scale(test)     # returns [0.0, 0.5, 1.0]
> ```

Using the original examples, the expected values for `test` are `[0.0, 0.010101, 0.020202]`, i.e., same as the one calculated for `train`.

Is this what you are asking?

**Comment by msluszniak** [comment]
Yeah, I agree that we need an interface that transforms data based on the previous fitting. I can transform all the functions into modules so they will have `fit` and `transform` functions and our functionality will be just fit and then transform calls on the same data.

**Comment by shaolang** [comment]
@msluszniak Yeah :) Are you making the changes, or should I do it, using @santiago-imelio pull request as a template? 'cos I'm new to the library, so if I were to do it, I might take slightly longer (not that I mind spending the time). :)

**Comment by msluszniak** [comment]
I'll make changes :))

**Comment by msluszniak** [comment]
I think that it will be problematic to implement efficiently Binarizer and OneHotEncoder. Currently, our OneHotEncoder works more like Binarizer and to efficiently translate labels into encoded ones we need a structure like a map which we don't have if we implement them in `Nx`.

**Comment by shaolang** [comment]
I thought the existing Binarizer implementation is fine, 'cos Binarizer doesn't need fitting?

The current one-hot encoder is simplistic, 'cos it imposes two prerequisites:
1. Only numerical values are acceptable ('cos Nx tensors doesn't support strings).
2. Number of classes must be known upfront; this also means:
    * There's no way to handle unknown/infrequent categories.
    * The function can't figure out/learn the number of categories based purely from training data.

These one-hot encoding prerequisites also means additional "preprocessing" must be done before the data is piped through Scholar's data pipeline. Moving the existing implementation to its own module opens the window to eliminate prerequisite number 2.

**Comment by msluszniak** [comment]
Yes, binarizer is fine, so moving it to a separate module will be only reasonable for consistency, but not required. The requirements for one-hot you mentioned are also valid. But removing additional preprocessing of data (label encoding) will require us to have a mapping from a current label which might be an arbitrary number into a 0-1 vector. At this moment I cannot figure out any better mapping that change both test and train using label encoding and then mapping is trivial (i -> ith vector).

**Comment by josevalim** [comment]
If we can't move them all, then my suggestion is to keep the functions in Scholar.Preprocessing. If they have a module shortcut, then we delegate to it, otherwise we keep the implementation as is.

**Comment by msluszniak** [comment]
We could move all of them, but fetching the value from mapping will be slow for label_encoder and one_hot encoder. We can implement mapping as {num_classes, 2} shape tensor and for each value make a linear scan and check the value of mapping. But I'm thinking if it's worth to have.

**Comment by shaolang** [comment]
One noob question: why Scholar doesn't accept Polars dataframes/lazyframes as inputs? That would open up a lot more options for creating the preprocessing pipeline. At the end of preprocessing, then and only then the transformation is converted to Nx tensors for the ML training/prediction.

Sorry if this question appears like I'm bashing the party, but I'm not. It really is a genuine question that I'm interested to hear the thoughts from the owners.

**Comment by josevalim** [comment]
You have the option to convert an Explorer.DataFrame to a tensor frame, which can then be regularly be used by Nx. Check Explorer docs.

**Comment by shaolang** [comment]
Thanks for pointing me to Explorer.

Are there plans to make Scholar work with Polars (via Explorer) for the Preprocessing and Impute modules natively? I'm coming from the perspective that scikit-learn works with NumPy and Pandas natively, making it easier to build the ML data pipeline.

That said, I am aware such integrations are more of a convenience and likely complicate Scholar's implementation/design.

**Comment by josevalim** [comment]
It really depends on what you want, but I imagine that a lot of the preprocessing can be done on Explorer too (and some is already possible, such as categorization). So it is a matter of adjusting expectations of where preprocessing happens.

Providing examples of what you want to do that feels clunky today will allow us to push work to either Explorer or Scholar to streamline the experience.

**Comment by josevalim** [comment]
And to be clear, the reason why there is a difference is because all Scholar code is JIT compiled and can also run on the GPU. This puts restrictions on which data formats it accepts.

**Comment by josevalim** [comment]
This is done. We need to decide if we want to keep the shortcut functions but that's a separate convo. :) Thank you @msluszniak !
